Web Inspector: add toggle for Site Isolation Debug Overlay in Engineering Settings

burg opened the pull request:
#### dba688ca0a87a68ccc2e850296935dd2c41a2b92
<pre>
Web Inspector: add toggle for Site Isolation Debug Overlay in Engineering Settings
<a href="https://bugs.webkit.org/show_bug.cgi?id=296076">https://bugs.webkit.org/show_bug.cgi?id=296076</a>
<a href="https://rdar.apple.com/155984834">rdar://155984834</a>

Reviewed by NOBODY (OOPS!).

Hook up the engineering UI setting to SetDeveloperPreferenceOverrides in UIProcess. The setting gets synced
to each frame/WebProcess and those processes are responsible for drawing overlays representing local frames.

* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/inspector/InspectorBackendClient.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideSetting):
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::SiteIsolationOverlay::drawRect):
Print the frameID, pid, and URL of each frame in the upper left corner.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
(WI.initializeTarget):
Restore developer override for backend target from frontend when initializing backend target.

* Source/WebInspectorUI/UserInterface/Base/Setting.js: Add new setting.

* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createEngineeringSettingsView):
Add new Site Isolation section to Engineering Settings.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setDeveloperPreferenceOverride):
Add new developer override.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dba688ca0a87a68ccc2e850296935dd2c41a2b92

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112173 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22382 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118250 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62605 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40468 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85214 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35977 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115120 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100928 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65644 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25286 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19064 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62095 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104687 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95366 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19141 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121576 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110787 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39247 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29193 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94034 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97169 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93857 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39086 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16876 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35275 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39135 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44623 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135018 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38770 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36307 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42107 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40513 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -632,6 +638,9 @@ WI.initializeTarget = function(target)
                 console.error(error);
             });
+
+        if (target.hasCommand("Page.overrideSetting"))
```
please include a `// COMPATIBILITY` comment

burg replied:
for what release? This command was added in iOS 13 according to our protocol snapshots.

dcrousso replied:
i meant for `InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay`, which i think you already added :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -499,6 +499,11 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
 
+        engineeringSettingsView.addSeparator();
+
+        let siteIsolationGroup = engineeringSettingsView.addGroup(WI.unlocalizedString("Site Isolation:"));
+        siteIsolationGroup.addSetting(WI.settings.engineeringShowSiteIsolationDebugOverlay, WI.unlocalizedString("Show Debug Overlay"));
```
i think this should only be shown if `InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay` exists

dcrousso replied:
i think you may have missed this change 😅

dcrousso commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
```diff
@@ -792,6 +793,19 @@ void WebInspectorUIProxy::setDeveloperPreferenceOverride(WebCore::InspectorBacke
+
+    case InspectorBackendClient::DeveloperPreference::SiteIsolationDebugOverlay:
+        if (RefPtr inspectedPage = m_inspectedPage.get()) {
+            uint32_t siteIsolationRegion = (uint32_t)DebugOverlayRegions::SiteIsolationRegion;
```
```suggestion
            auto siteIsolationRegion = static_cast<uint32_t>(DebugOverlayRegions::SiteIsolationRegion);
```

dcrousso replied:
actually is this even necessary?  could you just inline it instead?
```diff
 if (overrideValue && overrideValue.value())
-    debugOverlayRegions |= siteIsolationRegion;
+    debugOverlayRegions |= DebugOverlayRegions::SiteIsolationRegion;
 else
-    debugOverlayRegions &= ~siteIsolationRegion;
+    debugOverlayRegions &= ~DebugOverlayRegions::SiteIsolationRegion;
```

burg replied:
The `static_cast` has to stay there otherwise it doesn't know which integral type to use.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -237,6 +237,12 @@ WI.contentLoaded = function()
+    if (WI.engineeringSettingsAllowed()) {
+        WI.settings.engineeringShowSiteIsolationDebugOverlay.addEventListener(WI.Setting.Event.Changed, () => {
+            WI.mainTarget.PageAgent.overrideSetting(InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay, WI.settings.engineeringShowSiteIsolationDebugOverlay.value);
+        });
```
since Web Inspector uses its own event listener system, we try not to use arrow functions as they can keep more things alive than intended
```suggestion
        WI.settings.engineeringShowSiteIsolationDebugOverlay.addEventListener(WI.Setting.Event.Changed, function(event) {
            WI.mainTarget.PageAgent.overrideSetting(InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay, WI.settings.engineeringShowSiteIsolationDebugOverlay.value);
        });
}, WI.settings.engineeringShowSiteIsolationDebugOverlay);
```

dcrousso commented on Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
```diff
@@ -513,6 +514,10 @@ Inspector::Protocol::ErrorStringOr<void> InspectorPageAgent::overrideSetting(Ins
         return { };
 
+    case Inspector::Protocol::Page::Setting::SiteIsolationDebugOverlay:
+        m_client->setDeveloperPreferenceOverride(InspectorBackendClient::DeveloperPreference::SiteIsolationDebugOverlay, value);
```
should this also modify `inspectedPageSettings`?  or does it only have an effect in the UIProcess?

burg replied:
Due to site isolation, this needs to go out to UIProcess and get synced back to each web process, where the drawing actually happens.

burg committed:
Web Inspector: add toggle for Site Isolation Debug Overlay in Engineering Settings
https://bugs.webkit.org/show_bug.cgi?id=296076
rdar://155984834

Reviewed by NOBODY (OOPS!).

Hook up the engineering UI setting to SetDeveloperPreferenceOverrides in UIProcess. The setting gets synced
to each frame/WebProcess and those processes are responsible for drawing overlays representing local frames.

* Source/JavaScriptCore/inspector/protocol/Page.json:
* Source/WebCore/inspector/InspectorBackendClient.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideSetting):
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::SiteIsolationOverlay::drawRect):
Print the frameID, pid, and URL of each frame in the upper left corner.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
(WI.initializeTarget):
Restore developer override for backend target from frontend when initializing backend target.

* Source/WebInspectorUI/UserInterface/Base/Setting.js: Add new setting.

* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createEngineeringSettingsView):
Add new Site Isolation section to Engineering Settings.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setDeveloperPreferenceOverride):
Add new developer override.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -632,6 +638,10 @@ WI.initializeTarget = function(target)
             });
+
+        // COMPATIBILITY (iOS 18.0): Page.Setting.SiteIsolationDebugOverlay did not yet exist.
+        if (target.hasCommand("Page.overrideSetting") && InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay)
```
yeah i just meant `InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay`

as you pointed out, i think `Page.overrideSetting` exists for all supported versions, so im not sure that part of this check is necessary

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -237,6 +237,12 @@ WI.contentLoaded = function()
     WI.settings.experimentalAllowInspectingInspector.addEventListener(WI.Setting.Event.Changed, WI._allowInspectingInspectorSettingChanged, WI);
 
+    if (WI.engineeringSettingsAllowed()) {
+        WI.settings.engineeringShowSiteIsolationDebugOverlay.addEventListener(WI.Setting.Event.Changed, function() {
+            WI.mainTarget.PageAgent.overrideSetting(InspectorBackend.Enum.Page.Setting.SiteIsolationDebugOverlay, WI.settings.engineeringShowSiteIsolationDebugOverlay.value);
+        });
```
`WI.Object.prototype.addEventListener` will `console.assert()` if you dont provide a 3rd argument `thisObject` (which in this case id recommend being `WI.settings.engineeringShowSiteIsolationDebugOverlay` since that's the thing being listened for and to match what's done elsewhere)
